Tests/NIOHTTP1Tests/HTTPDecoderTest.swift

```diff
@@ -0,0 +1,192 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import XCTest
+import Dispatch
+import NIO
+import NIOHTTP1
+
+class HTTPDecoderTest: XCTestCase {
+    private var channel: EmbeddedChannel!
+    private var loop: EmbeddedEventLoop!
+
+    override func setUp() {
+        self.channel = EmbeddedChannel()
+        self.loop = channel.eventLoop as! EmbeddedEventLoop
+    }
+
+    override func tearDown() {
+        self.channel = nil
+        self.loop = nil
+    }
+
+    func testDoesNotDecodeRealHTTP09Request() throws {
+        XCTAssertNoThrow(try channel.pipeline.add(handler: HTTPRequestDecoder()).wait())
+
+        // This is an invalid HTTP/0.9 simple request (too many CRLFs), but we need to
+        // trigger https://github.com/nodejs/http-parser/issues/386 or http_parser won't
+        // actually parse this at all.
+        var buffer = channel.allocator.buffer(capacity: 64)
+        buffer.write(staticString: "GET /a-file\r\n\r\n")
+
+        do {
+            try channel.writeInbound(buffer)
+            XCTFail("Did not error")
+        } catch HTTPParserError.invalidVersion {
+            // ok
+        } catch {
+            XCTFail("Unexpected error: \(error)")
+        }
+
+       loop.run()
+    }
+
+    func testDoesNotDecodeFakeHTTP09Request() throws {
+        XCTAssertNoThrow(try channel.pipeline.add(handler: HTTPRequestDecoder()).wait())
+
+        // This is a HTTP/1.1-formatted request that claims to be HTTP/0.9.
+        var buffer = channel.allocator.buffer(capacity: 64)
+        buffer.write(staticString: "GET / HTTP/0.9\r\nHost: whatever\r\n\r\n")
+
+        do {
+            try channel.writeInbound(buffer)
+            XCTFail("Did not error")
+        } catch HTTPParserError.invalidVersion {
+            // ok
+        } catch {
+            XCTFail("Unexpected error: \(error)")
+        }
+
+        loop.run()
+    }
+
+    func testDoesNotDecodeHTTP2XRequest() throws {
+        XCTAssertNoThrow(try channel.pipeline.add(handler: HTTPRequestDecoder()).wait())
+
+        // This is a hypothetical HTTP/2.0 protocol request, assuming it is
+        // byte for byte identical (which such a protocol would never be).
+        var buffer = channel.allocator.buffer(capacity: 64)
+        buffer.write(staticString: "GET / HTTP/2.0\r\nHost: whatever\r\n\r\n")
+
+        do {
+            try channel.writeInbound(buffer)
+            XCTFail("Did not error")
+        } catch HTTPParserError.invalidVersion {
+            // ok
+        } catch {
+            XCTFail("Unexpected error: \(error)")
+        }
+
+        loop.run()
+    }
+
+    func testToleratesHTTP13Request() throws {
+        XCTAssertNoThrow(try channel.pipeline.add(handler: HTTPRequestDecoder()).wait())
+
+        // We tolerate higher versions of HTTP/1 than we know about because RFC 7230
+        // says that these should be treated like HTTP/1.1 by our users.
```

Comment:
omg, what's wrong with RFC7230?

Reply:
I mean, the right thing for our users to do is to send HTTP/1.1 responses in response to HTTP/1.2 requests. But we can't really enforce that sensibly.

Reply:
Anyway it's largely theoretical as HTTP/1.2 is never gonna happen.

![](https://media.vanityfair.com/photos/5630dd7e5350044647455930/master/w_900,c_limit/tumblr_inline_nwsolxofKo1s7ww91_500.gif)